// aerobatic-cli/commands/clientip.js
const chalk = require('chalk');
const _ = require('lodash');
const urlJoin = require('url-join');
const publicIp = require('public-ip');
const output = require('../lib/output');
const api = require('../lib/api');

module.exports = program => {
  // If a value argument is specified, then set clientIP ranges
  if (program.delete === true) {
    return deleteClientIpRange(program);
  }
  if (program.value) {
    return setClientIpRange(program);
  }

  return display(program);
};

function printRange(range) {
  return range.join(', ');
}

function display(program) {
  const {clientIpRange} = program.website;
  if (!_.isObject(clientIpRange) || _.isEmpty(clientIpRange)) {
    return output(
      chalk.yellow('No client IP range restrictions have been set.')
    );
  }

  output.blankLine();
  output('Website access is limited to client IPs falling in the ranges:');
  output.blankLine();

  const stages = Object.keys(clientIpRange);
  if (stages.length === 1 && stages[0] === '_global') {
    output(`All deploy stages: ${printRange(clientIpRange._global)}`);
  } else {
    _.forEach(_.omit(clientIpRange, '_global'), (value, stage) => {
      output(`${chalk.dim(_.padStart(stage, 20))}: ${printRange(value)}`);
    });
    if (clientIpRange._global) {
      output(
        `${chalk.dim(_.padStart('All other stages', 20))}: ${printRange(
          clientIpRange._global
        )}`
      );
    }
  }
  output.blankLine();
}

function putValueWithApi(program) {
  const rangeValue = _.map(program.value.split(','), _.trim);

  return api
    .put({
      url: urlJoin(
        _.compact([
          program.apiUrl,
          `/apps/${program.website.appId}/clientip-range`,
          program.stage
        ])
      ),
      authToken: program.authToken,
      body: {
        clientIpRange: rangeValue
      }
    })
    .then(() => {
      let message;
      if (program.stage) {
        message = `Allowed client IP ranges updated to ${printRange(
          rangeValue
        )} for stage ${program.stage}`;
      } else {
        message = `Allowed client IP ranges updated to ${printRange(
          rangeValue
        )}`;
      }

      output(chalk.green(message));
      output.blankLine();
    });
}

function setClientIpRange(program) {
  if (!_.isString(program.value)) {
    return Promise.reject(
      Error.create(
        'Must provide comma delimited list of IPs or the value "myip" in the --value option',
        {
          formatted: true
        }
      )
    );
  }

  output.blankLine();
  if (program.value === 'myip') {
    publicIp.v4().then(ip => {
      program.value = ip;
      return putValueWithApi(program);
    });
  } else {
    return putValueWithApi(program);
  }
}

function deleteClientIpRange(program) {
  return api
    .delete({
      url: urlJoin(
        _.compact([
          program.apiUrl,
          `/apps/${program.website.appId}/clientip-range`,
          program.stage
        ])
      ),
      authToken: program.authToken
    })
    .then(() => {
      let message;
      if (program.stage) {
        message = `Client IP range restrictions have been cleared for stage ${
          program.stage
        }`;
      } else {
        message = `The default client IP range restrictions have been cleared. Any stage-specific settings remain intact. Run ${output.command(
          'aero clientip'
        )} to see the effective values.`;
      }

      output(chalk.green(message));
    });
}
